plugins/c-ares-dns-client.c

```diff
@@ -0,0 +1,757 @@
+
+	if (!init_ares_watch(request, s, read, write))
+		ofono_error("%s: couldn't create ares_watch for socket %d",
+			      __func__, s);
+}
+
+/*
+ * Converts a c-ares status code to the corresponding dns_client status code.
+ * We do this to completely encapsulate c-ares. In theory, we should be able to
+ * replace it with a different asynchronous DNS library without changing our
+ * clients.
+ */
+static ofono_dns_client_status_t status_from_ares_status(int ares_status)
+{
+	switch(ares_status) {
+	case ARES_SUCCESS:
+		return OFONO_DNS_CLIENT_SUCCESS;
+	case ARES_ENODATA:
+		return OFONO_DNS_CLIENT_ERROR_NO_DATA;
+	case ARES_EFORMERR:
+		return OFONO_DNS_CLIENT_ERROR_FORM_ERR;
+	case ARES_ESERVFAIL:
+		return OFONO_DNS_CLIENT_ERROR_SERVER_FAIL;
+	case ARES_ENOTFOUND:
+		return OFONO_DNS_CLIENT_ERROR_NOT_FOUND;
+	case ARES_ENOTIMP:
+		return OFONO_DNS_CLIENT_ERROR_NOT_IMP;
+	case ARES_EREFUSED:
+		return OFONO_DNS_CLIENT_ERROR_REFUSED;
+	case ARES_EBADQUERY:
+	case ARES_EBADNAME:
+	case ARES_EBADFAMILY:
+	case ARES_EBADRESP:
+		return OFONO_DNS_CLIENT_ERROR_BAD_QUERY;
+	case ARES_ECONNREFUSED:
+		return OFONO_DNS_CLIENT_ERROR_NET_REFUSED;
+	case ARES_ETIMEOUT:
+		return OFONO_DNS_CLIENT_ERROR_TIMED_OUT;
+	default:
+		return OFONO_DNS_CLIENT_ERROR_UNKNOWN;
+	}
+}
+
+/*
+ * Callback that is invoked by c-ares when an asynchronous name resolution
+ * request that we have previously initiated is complete.
+ */
+static void ares_request_cb(void *arg, int ares_status, int timeouts,
+			    struct hostent *hostent)
+{
+	struct sockaddr_in sin;
+	struct sockaddr_in6 sin6;
+	int addr_length;
+	void *addr_buffer;
+	char ip_addr_string[INET6_ADDRSTRLEN];
+	struct sockaddr *ip_addr;
+	struct ares_request *request = (struct ares_request *)arg;
+
+	DBG("");
+
+	if (!request->running)
+		return;
+
+	/* Stop the request. It will be deleted later from the idle loop. */
+	stop_ares_request(request);
+
+	if (ares_status != ARES_SUCCESS) {
+		DBG("ares request for '%s' failed: %s",
+				request->hostname, ares_strerror(ares_status));
+		/* Notify client. */
+		request->cb(request->data, status_from_ares_status(ares_status),
+			    NULL);
+		return;
+	}
+
+	if (hostent->h_addrtype != AF_INET && hostent->h_addrtype != AF_INET6) {
+		DBG("unsupported addrtype: %d",
+				hostent->h_addrtype);
+		request->cb(request->data, OFONO_DNS_CLIENT_ERROR_NO_DATA,
+			    NULL);
+		return;
+	}
+
+	if (hostent->h_addrtype == AF_INET) {
+		memset(&sin, 0, sizeof(sin));
+		sin.sin_family = AF_INET;
+		addr_length = sizeof(sin.sin_addr.s_addr);
+		addr_buffer = &sin.sin_addr.s_addr;
+		ip_addr = (struct sockaddr *) &sin;
+	} else {  /* AF_INET6 */
+		memset(&sin6, 0, sizeof(sin6));
+		sin6.sin6_family = AF_INET6;
+		addr_length = sizeof(sin6.sin6_addr.s6_addr);
+		addr_buffer = &sin6.sin6_addr.s6_addr;
+		ip_addr = (struct sockaddr *) &sin6;
+	}
+
+	if (hostent->h_length > addr_length) {
+		DBG("address too large: %u bytes",
```

Comment:
The CB prints the error too. I have changed those which returned OFONO_DNS_CLIENT_ERROR_NO_DATA as that one can happen for different conditions.